test/test.go

```diff
@@ -0,0 +1,213 @@
+	// read dirs in base tests dir
+	testDirs, err := ioutil.ReadDir(testsRootDir)
+	if err != nil {
+		return nil, err
+	}
+
+	// find test results in each dir, skip if invalid test dir
+	for _, testDir := range testDirs {
+		testDirPath := filepath.Join(testsRootDir, testDir.Name())
+
+		// read one level of file set only <root_tests_dir>/<test_dir>/file_to_get
+		testFiles, err := filepath.Glob(filepath.Join(testDirPath, "*"))
+		if err != nil {
+			return nil, err
+		}
+
+		// find step-info in dir
+		var stepInfo *models.TestResultStepInfo
+		for _, file := range testFiles {
+			if strings.HasSuffix(file, "step-info.json") {
+				stepInfoFileContent, err := fileutil.ReadBytesFromFile(file)
+				if err != nil {
+					return nil, err
+				}
+				if err := json.Unmarshal(stepInfoFileContent, &stepInfo); err != nil {
+					return nil, err
+				}
+			}
+		}
+
+		// if no step-info.json then skip
+		if stepInfo == nil {
+			continue
+		}
+
+		// get the handler that can manage test type contained in the dir
+		for _, handler := range converters.List() {
+			handler.SetFiles(testFiles)
+			// skip if couldn't find handler for content type
+			if handler.Detect() {
+				xml, err := handler.XML()
+				if err != nil {
+					return nil, err
+				}
+
+				// so here I will have image paths, xml data, and step info per test dir in a bundle info
+				bundle = append(bundle, ResultAssets{
+					XMLContent: xml,
+					ImagePaths: findImages(filepath.Join(testsRootDir, testDir.Name())),
+					StepInfo:   *stepInfo,
+				})
+			}
+		}
+	}
+	return bundle, nil
+}
+
+func checkResponse(resp *http.Response) error {
+	if resp.StatusCode < 200 || 299 < resp.StatusCode {
+		bodyData, err := ioutil.ReadAll(resp.Body)
+		if err != nil {
+			return fmt.Errorf("unsuccessful response code: %d and failed to read body, error: %s", resp.StatusCode, err)
+		}
+		return fmt.Errorf("unsuccessful response code: %d\nbody:\n%s", resp.StatusCode, bodyData)
+	}
+	return nil
+}
+
+// Upload ...
+func (bundle Bundle) Upload(client *http.Client, endpointURL string) error {
+	// generate UUID->filename and UUID->filepath maps and fill them with the assets need to be uploaded
+	assetUploadRequestMap, assetIDPathMap := map[string]string{}, map[string]string{}
+	for _, testResultAsset := range bundle {
+		for _, path := range testResultAsset.ImagePaths {
+			assetUUID := uuid.New().String()
+			assetUploadRequestMap[assetUUID] = filepath.Base(path)
+			assetIDPathMap[assetUUID] = path
+		}
+	}
+
+	// Get an uploadURL map from the server UUID->URL to upload
+	var assetUploadRequestMapData bytes.Buffer
+	if err := json.NewEncoder(&assetUploadRequestMapData).Encode(assetUploadRequestMap); err != nil {
+		return err
+	}
+
+	req, err := http.NewRequest(http.MethodGet, endpointURL, bytes.NewReader(assetUploadRequestMapData.Bytes()))
+	if err != nil {
+		return err
+	}
+
+	resp, err := client.Do(req)
+	if err != nil {
+		return err
+	}
+
+	defer func() {
```

Comment:
I think this will shadow the other returned errors.

You should use a different error variable `cerr` like this:
```
defer func() {
		if cerr := response.Body.Close(); cerr != nil {
			log.Warnf("Failed to close response body, error: %s", cerr)
		}
	}()
```

Reply:
by this sample code it looks like it works a little bit differently than the expected:
```
package main

import "fmt"

func test() (err error) {
	err = fmt.Errorf("tomi")

	defer func() {
		if err := fmt.Errorf("nemtomi"); err != nil {
			fmt.Println("from defer:", err)
		}
	}()

	fmt.Println("outer scope:", err)
	return
}

func main() {
	fmt.Println("main:", test())
}
```